### PR TITLE
OCLOMRS-958:Errors from the backend should marked the field that caused the error

### PR DESCRIPTION
**Pull request by jwnasambu**
# JIRA TICKET NAME:
[Errors from the backend should marked the field that caused the error](https://issues.openmrs.org/browse/OCLOMRS-958)

# Summary:

### CODE REVIEW

**Comment by hadijahkyampeire** [comment]
@jwnasambu why did you change this PR to a draft? what's up with it?

**Comment by jwnasambu** [comment]
@dkayiwa, @ibacher  Kindly how do I solve this persistent error `ERROR: Service 'solr' failed to build: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit

The command "bash ./start_local_instance.sh" exited with 1. ` causing the travis failure?

**Comment by jwnasambu** [comment]
@dkayiwa thanks for the help. @hadijahkyampeire the ticket is not complete that is why its in `draft` state

**Comment by dkayiwa** [comment]
@jwnasambu is the pull rate limit error still a problem?

**Comment by jwnasambu** [comment]
@dkayiwa Am still fixing the issue but I will gladly communicate once I have pushed the changes. Thanks for asking.

**Comment by dkayiwa** [comment]
@jwnasambu the issue of pull rate limit?

**Comment by jwnasambu** [comment]
@dkayiwa after you restarted the server, the issue of pull rate limit cease to be a problem for now. Thanks

**Comment by ibacher** [comment]
@jwnasambu This looks good so far. Can you rebase your branch on the current `master` (which should fix the test failures). The last thing this PR needs is this (from the ticket):

> 3. We should update the form so that anything in the new error object `__details__` field gets displayed at the bottom of the form, where errors from the backend are displayed today.

File: src/redux/utils.ts [hunk 1]
```diff
@@ -114,15 +114,17 @@ export function errorMsgResponse(response: any) {
...
         Array.isArray(response.data[key])
           ? response.data[key].join(",")
           : response.data[key]
       );
     }
   }
-  return errorMsgResponse.length > 0
-    ? errorMsgResponse.join("\n")
-    : genericErrorMessage;
+  errorMsgResponse[key] =
```
Comment (ibacher) [inline src/redux/utils.ts:124]: This line should be moved before line 118.

File: src/redux/utils.ts [hunk 2]
```diff
@@ -114,15 +114,17 @@ export function errorMsgResponse(response: any) {
...
+  Array.isArray(response.data[key])
+    ? response.data[key].join(",")
+    : response.data[key];
```
Comment (ibacher) [inline src/redux/utils.ts:127]:
```suggestion
```

File: src/redux/utils.ts [hunk 3]
```diff
@@ -114,15 +114,17 @@ export function errorMsgResponse(response: any) {
   } else {
     for (let key in response.data) {
       errorMsgResponse.push(
+        errorMsgResponse["__detail__"] = response.data.detail,
```
Comment (ibacher) [inline src/redux/utils.ts:117]: This line should be outside the for loop.

Comment (ibacher) [reply src/redux/utils.ts:117]: Also, we should probably get rid of the `if`...`else` altogether. Maybe something like:

```ts
if (response.data && Object.prototype.hasOwnProperty.call(response.data, "detail")) { 
  errorMsgResponse["__detail__"] = response.data.detail;
}

for (let key in response.data) {
  if (key === "__detail__") continue;
  errorMsgResponse[key] =
    // code goes here
}
```

Comment (ibacher) [reply src/redux/utils.ts:117]: We should also change the type of `errorMsgResponse` to:

```ts
const errorMsgResponse: {[key: string]: string};
```

**Review by ibacher — CHANGES_REQUESTED** [review]

File: src/redux/utils.ts [hunk 4]
```diff
@@ -104,26 +104,17 @@ export function invalidateCache(action: string, dispatch: Function) {
 
 export function errorMsgResponse(response: any) {
   let errorMsgResponse: string[] = [];
```
Comment (ibacher) [inline src/redux/utils.ts:106]:
```suggestion
```

File: src/redux/utils.ts [hunk 5]
```diff
@@ -104,26 +104,17 @@ export function invalidateCache(action: string, dispatch: Function) {
...
-    errorMsgResponse.push(response.data.detail);
-  } else {
-    for (let key in response.data) {
-      errorMsgResponse.push(
-        Array.isArray(response.data[key])
-          ? response.data[key].join(",")
-          : response.data[key]
-      );
-    }
+  const errorMsgResponse: {[key: string]: string} = {}; = "Action could not be completed. Please retry.";
```
Comment (ibacher) [inline src/redux/utils.ts:107]:
```suggestion
  const errorMsgResponse: {[key: string]: string} = {
    "__detail__": "Action could not be completed. Please retry."
  };
```

File: src/redux/utils.ts [hunk 6]
```diff
@@ -104,26 +104,17 @@ export function invalidateCache(action: string, dispatch: Function) {
...
+
+  if (response.data && Object.prototype.hasOwnProperty.call(response.data, "detail")) { 
+    errorMsgResponse["__detail__"] = response.data.detail;
+  }
+  
+  for (let key in response.data) {
+    if (key === "__detail__") continue;
+    errorMsgResponse[key] =
+      // code goes here
```
Comment (ibacher) [inline src/redux/utils.ts:116]: This comment was meant as a place holder for something like:

```ts
Array.isArray(response.data[key])
 ? response.data[key].join(",")
 : response.data[key]
```

**Comment by jwnasambu** [comment]
@ibacher locally the changes build successfully but when I try to rebase everything flops and am forced to begin a fresh.  Is it okay for me to close this branch and begin a fresh?

**Pull request closed without merging**